File: languages/go/client/authenticity.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/authenticity"
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/entity/record"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// AuthenticityClient represents a client for interacting with the [Bloock Authenticity service].
//
// [Bloock Authenticity service]: https://bloock.com
type AuthenticityClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewAuthenticityClient creates a new instance of the AuthenticityClient with default configuration.
func NewAuthenticityClient() AuthenticityClient {
	return AuthenticityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewAuthenticityClientWithConfig creates a new instance of the AuthenticityClient with the provided configuration.
func NewAuthenticityClientWithConfig(configData *proto.ConfigData) AuthenticityClient {
	return AuthenticityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// GenerateEcdsaKeys generates ECDSA key pair for signing records.
// Deprecated: Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
func (c *AuthenticityClient) GenerateEcdsaKeys() (key.KeyPair, error) {
	res, err := c.bridgeClient.Key().GenerateLocalKey(context.Background(), &proto.GenerateLocalKeyRequest{
		ConfigData: c.configData,
		KeyType:    key.KeyTypeToProto[key.EcP256k],
	})

	if err != nil {
		return key.KeyPair{}, err
	}

	if res.Error != nil {
		return key.KeyPair{}, errors.New(res.Error.Message)
	}

	return key.NewEcdsaKeysFromProto(res), nil
}

// Sign signs a Bloock record using the specified signer.
func (c *AuthenticityClient) Sign(r record.Record, s authenticity.Signer) (authenticity.Signature, error) {
	res, err := c.bridgeClient.Authenticity().Sign(context.Background(), &proto.SignRequest{
		ConfigData: c.configData,
		Record:     r.ToProto(),
		Signer:     s.ToProto(),
	})

	if err != nil {
		return authenticity.Signature{}, err
	}

	if res.Error != nil {
		return authenticity.Signature{}, errors.New(res.Error.Message)
	}

	return authenticity.NewSignatureFromProto(res.Signature), nil
}

// Verify verifies the authenticity of a Bloock record.
func (c *AuthenticityClient) Verify(r record.Record) (bool, error) {
	res, err := c.bridgeClient.Authenticity().Verify(context.Background(), &proto.VerifyRequest{
		ConfigData: c.configData,
		Record:     r.ToProto(),
	})

	if err != nil {
		return false, err
	}

	if res.Error != nil {
		return false, errors.New(res.Error.Message)
	}

	return res.Valid, nil
}

// GetSignatures retrieves the signatures associated with a Bloock record.
func (c *AuthenticityClient) GetSignatures(r record.Record) ([]authenticity.Signature, error) {
	bridgeClient := bridge.NewBloockBridge()
	res, err := bridgeClient.Authenticity().GetSignatures(context.Background(), &proto.GetSignaturesRequest{ConfigData: c.configData, Record: r.ToProto()})

	if err != nil {
		return []authenticity.Signature{}, err
	}

	if res.Error != nil {
		return []authenticity.Signature{}, errors.New(res.Error.Message)
	}

	signatures := make([]authenticity.Signature, len(res.Signatures))
	for i, signature := range res.Signatures {
		signatures[i] = authenticity.NewSignatureFromProto(signature)
	}
	return signatures, nil
}

File: languages/js/src/client/authenticity.ts
import { BloockBridge } from "../bridge/bridge";
import { ConfigData } from "../bridge/proto/config";
import { EcdsaKeyPair, KeyPair } from "../entity/key";
import { NewConfigData } from "../config/config";
import { Signature } from "../entity/authenticity";
import { Record } from "../entity/record";
import {
  GetSignaturesRequest,
  SignRequest,
  VerifyRequest
} from "../bridge/proto/authenticity";
import { Signer } from "../entity/authenticity/signer";
import { GenerateLocalKeyRequest } from "../bridge/proto/keys";
import { KeyType } from "../bridge/proto/keys_entities";

/**
 * Represents a client for interacting with the [Bloock Authenticity service](https://dashboard.bloock.com/login).
 */
export class AuthenticityClient {
  private bridge: BloockBridge;
  private configData: ConfigData;

  /**
   * Creates a new instance of the AuthenticityClient with default configuration.
   * @param configData 
   */
  constructor(configData?: ConfigData) {
    this.bridge = new BloockBridge();
    this.configData = NewConfigData(configData);
  }

  /**
   * Generates ECDSA key pair for signing records.
   * @returns {Promise<KeyPair>} An object containing both the public and the private key
   * @deprecated Will be deleted in future versions. Use KeyClient.newLocalKey function instead.
   */
  public async generateEcdsaKeyPair(): Promise<KeyPair> {
    let request = GenerateLocalKeyRequest.fromPartial({
      configData: this.configData,
      keyType: KeyType.EcP256k
    });

    return this.bridge
      .getKey()
      .GenerateLocalKey(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return EcdsaKeyPair.fromProto(res);
      });
  }

  /**
   * Signs a Bloock record using the specified signer.
   * @param record 
   * @param signer 
   * @returns 
   */
  public async sign(record: Record, signer: Signer): Promise<Signature> {
    return this.bridge
      .getAuthenticity()
      .Sign(
        SignRequest.fromPartial({
          configData: this.configData,
          record: record.toProto(),
          signer: signer.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Signature.fromProto(res.signature!);
      });
  }

  /**
   * Verifies the authenticity of a Bloock record.
   * @param record 
   * @returns 
   */
  public async verify(record: Record): Promise<boolean> {
    return this.bridge
      .getAuthenticity()
      .Verify(
        VerifyRequest.fromPartial({
          configData: this.configData,
          record: record.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.valid;
      });
  }

  /**
   * Gets the signatures associated with a Bloock record.
   * @param record 
   * @returns 
   */
  async getSignatures(record: Record): Promise<Signature[]> {
    return this.bridge
      .getAuthenticity()
      .GetSignatures(
        GetSignaturesRequest.fromPartial({
          configData: this.configData,
          record: record.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.signatures.map(x => Signature.fromProto(x));
      });
  }
}
